refactor: Check translatable format strings at compile-time

maflcko opened the pull request:
All translatable format strings are fixed. This change surfaces errors in them at compile-time.

The implementation achieves this by allowing to delay the translation (or `std::string` construction) that previously happened in `_()` by returning a new type from this function. The new type can be converted to `bilingual_str` where needed.

This can be tested by adding a format string error in an original string literal and observing a new compile-time failure.

Fixes https://github.com/bitcoin/bitcoin/issues/30530

Commit:
refactor: Delay translation of _() literals

This is required for a future commit that requires _() to be consteval
for format literals.

Co-Authored-By: Ryan Ofsky <ryan@ofsky.org>

Commit:
refactor: Check translatable format strings at compile-time

ryanofsky reviewed (approved):
Code review ACK fabb28f6c8f772d38917a23dfbda706292070ba1. This looks great, and implementation is very clean. However, I think it could be simplified significantly if there were a scripted diff commit to replace `strprintf(Untranslated("..."), ...)` with `Untranslated(strprintf("...", ...))`, which would be a good change on its own for clarity and consistency. More details in comments below.

stickies-v commented on src/util/translation.h:
```diff
@@ -6,10 +6,43 @@
+    return G_TRANSLATION_FUN ? G_TRANSLATION_FUN(lit) : lit;
+}
+/** Type to denote whether an orginal string literal is translatable */
+template <bool translatable = true>
```
nit: I would prefer templating this on an enum `{Translateable,Untranslateable}` for readability over the non-obvious `<{true,false}>` boolean

ryanofsky replied:
> nit: I would prefer templating this on an enum `{Translateable,Untranslateable}` for readability over the non-obvious `<{true,false}>` boolean

Note: with changes in suggested branch it should no longer be necessary for this to be templated.

stickies-v reviewed (commented):
Concept ACK, and this is a smaller diff than I'd expected to implement this, nice!

I need to think about it more and work out some code, but I feel like the new interface is not as intuitive yet as it could be. Will follow up with that in next review.

ryanofsky commented on src/util/translation.h:
```diff
@@ -22,6 +41,13 @@ struct bilingual_str {
     std::string original;
     std::string translated;
 
+    bilingual_str() = default;
+    bilingual_str(std::string original, std::string translated) : original{original}, translated{translated} {}
+    template <bool translatable>
+    bilingual_str(util::Original<translatable> original) : original{original.lit}, translated{original.translate()}
+    {
+    }
```
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1793642022

Yes I do think that's better and basically made the same change in my branch, though without the str() method since it seems it makes it a little less clear what return type is.

ryanofsky commented on src/util/translation.h:
```diff
@@ -6,10 +6,43 @@
+    consteval bilingual_fmt(Original<translatable> o) : original{o.lit}, translatable{translatable}
+    {
+    }
+    std::string translate() { return translate(original.fmt); }
```
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1793744937

This is a good catch, but it seems better to just drop the template method instead of trying to fix it up and use it. It just seems to add complexity and it doesn't seem like there is another use-case for it.

ryanofsky commented on src/qt/walletcontroller.cpp:
```diff
@@ -431,7 +431,7 @@ void RestoreWalletActivity::finish()
     } else {
         //: Title of message box which is displayed when the wallet is successfully restored.
-        QMessageBox::information(m_parent_widget, tr("Restore wallet message"), QString::fromStdString(Untranslated("Wallet restored successfully \n").translated));
+        QMessageBox::information(m_parent_widget, tr("Restore wallet message"), QString::fromStdString(Untranslated("Wallet restored successfully \n").translate()));
```
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1793707636

Would be reasonable to change this but no more change is actually required here with my suggested branch.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1793707636

Simplified this further in 6fb39655c5cd2bebad902d271d1edae64b817d2b from #31072. It might just be a bug that this string is not translated, and there also seems to be trailing whitespace in the string. Current commit just simplifies code and keeps existing behavior though.

ryanofsky commented on src/util/translation.h:
```diff
@@ -7,12 +7,31 @@
+}
+/** Type to denote whether an orginal string literal is translatable */
+template <bool translatable = true>
+struct Original {
```
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1793488009

> in [aaaa4fb](https://github.com/bitcoin/bitcoin/commit/aaaa4fb20156b4375d92e1eca4acc90a425a1896): seems more straightforward to just extend `ConstevalStringLiteral` instead of creating a new `util::Original`? This would 1) minimize the diff and 2) I find `ConstevalStringLiteral` a much more helpful name than `Original`.

I think this is only a question of naming since `Original` class is replacing `ConstevalStringLiteral` this commit. In by branch I called this ConstEvalTranslated, since it's the type name for values returned from the underscore function, and is not just referring to a generic string literal. But I didn't put much thought into the name and something else may be better.

ryanofsky commented on src/util/translation.h:
```diff
@@ -47,8 +48,30 @@ inline bilingual_str operator+(bilingual_str lhs, const bilingual_str& rhs)
+};
+} // namespace util
+
+consteval auto _(util::Translatable<true> str) { return str; }
```
In commit "refactor: Delay translation of Untranslated() or _() literals" (fa62737ae398bd56ef9e68b8d0a4d217cfcc532b)

Thinking about this more it seems like if `_` function does not be consteval, and could just a template type that inherits from `bilingual_str`, and included the number of format specifiers included in the string in its type information. This would be nice because runtime semantics of `_` would be unchanged, it would still do the translation at time it is called, and there would be no need to make any `translated` -> `translate()` replacements. 

I think the main downside to this approach is would be need to refactor CheckNumFormatSpecifiers but there would be fewer changes overall. The other downside would be that _ would need to be aware the translated strings might be a format string and count the specifiers instead of just passing it along blindly. Neither downside would have to exist if c++ just allowed arguments to consteval functions to be treated as compile time constants, which they are by definition, but they did not do this (I think because stylistically they want to see compile-time arguments passed as `<>` arguments not `()` arguments.)

Not sure about any of this, though, would need to experiment.

ryanofsky reviewed (approved):
Code review ACK fae6acacc0216aa07f5a4a0ee75fbade620d7023. This is a nice improvement overall, and definitely better than the status quo due to extra type checking it provides. I do think there is extra complexity here and extra changes that could go away if we adopt #31072 and stop passing untranslated format strings to strprintf. 

I also think it probably would be nicer if `_()` would continue to look up translations right away, instead of delaying them until conversion to bilingual str. I think it could return a bilingual_str with extra compile-time information instead of returning different type with `.lit` and `.translate()` fields. I'm not sure this is possible though, and it's likely there's something I'm overlooking.

EDIT: I am implemented an approach in #31074 that makes `_()` and `Untranslated()` both return `bilingual_str` with `.original` and `.translated` fields and no not delay the time of translation. Unfortunately, because C++ does not currently allow marking function parameters `constexpr`, it means strings needs to be passed as template parameters instead of function parameters, resulting in a messier syntax, even though the implementation and API are simpler. I closed the PR because it is not a clear win over this one.

ryanofsky commented on src/common/init.cpp:
```diff
@@ -107,7 +107,7 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
         }
     } catch (const std::exception& e) {
-        return ConfigError{ConfigStatus::FAILED, Untranslated(e.what())};
+        return ConfigError{ConfigStatus::FAILED, Untranslated(std::string{e.what()})};
```
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1793837788

Opened #31072 to implement suggested simplification. Marking this thread resolved since it it's not meant to block this PR in any case.

l0rinc commented:
Concept ACK, nice to see these verifications being done earlier

ryanofsky reviewed (approved):
Code review ACK 2c15a858ca7b3f14eaebf4b39134cd0e9c559273. No changes since the last review other than rebase.

I do think it would simplify this PR and make review more meaningful if #31072 could be merged first. This PR is combining application code changes which require looking at surrounding code and seeing if the changes fit in, with translation class changes which require thinking about API design. I think it's harder to evaluate both things clearly when they are bundled in the same PR. I wouldn't object to this PR going in first though since it has been open longer.

ryanofsky reviewed (approved):
Code review ACK f4df783f1ca22d96476d52ec5d1929547691ba13. Just rebased and reordered commit since last review

l0rinc reviewed (commented):
utACK f4df783f1ca22d96476d52ec5d1929547691ba13

hodlinator reviewed (approved):
ACK f4df783f1ca22d96476d52ec5d1929547691ba13

Increased validation of format strings. :+1: 

---

Maybe amend PR summary or final commit by spelling out something like:

Delays `bilingual_str`-construction until the end of the overload of `tinyformat::format` in *translation.h*. The `fmt` parameter to the overload is changed from `bilingual_str` -> `util::bilingual_fmt`, the latter being a new type which contains a `ConstevalFormatString`. `bilingual_fmt` in turn is only constructible from the new `util::Translatable` type, which is only returned by calling one of:
- `Translatable<true> _(const char*)`
- `Translatable<false> Untranslated(const char*)`

---

Iff you re-touch, consider making code a bit less template-heavy by splitting `Translatable` in two:

```diff
diff --git a/src/util/translation.h b/src/util/translation.h
index 2da4f68e55..657cd21f9c 100644
--- a/src/util/translation.h
+++ b/src/util/translation.h
@@ -58,31 +58,32 @@ inline std::string translate(const char* lit)
 {
     return G_TRANSLATION_FUN ? G_TRANSLATION_FUN(lit) : lit;
 }
-/** Type to denote whether an original string literal is translatable */
-template <bool translatable = true>
 struct Translatable {
     const char* const lit;
     consteval Translatable(const char* str) : lit{str} { assert(lit); }
-    std::string translate() const { return translatable ? util::translate(lit) : lit; }
+    std::string translate() const { return util::translate(lit); }
     operator bilingual_str() const { return {lit, translate()}; }
 };
+struct Untranslatable {
+    const char* const lit;
+    consteval Untranslatable(const char* str) : lit{str} { assert(lit); }
+    operator bilingual_str() const { return {lit, lit}; }
+};
 template <unsigned num_params>
 struct bilingual_fmt {
     const ConstevalFormatString<num_params> original;
     const bool translatable;
-    template <bool translatable>
-    consteval bilingual_fmt(Translatable<translatable> o) : original{o.lit}, translatable{translatable}
-    {
-    }
+    consteval bilingual_fmt(Translatable o) : original{o.lit}, translatable{true} {}
+    consteval bilingual_fmt(Untranslatable o) : original{o.lit}, translatable{false} {}
     std::string translate() { return translatable ? util::translate(original.fmt) : original.fmt; }
 };
 } // namespace util

-consteval auto _(util::Translatable<true> str) { return str; }
+consteval auto _(util::Translatable str) { return str; }

 /** Mark a bilingual_str as untranslated */
 inline bilingual_str Untranslated(std::string original) { return {original, original}; }
-consteval util::Translatable<false> Untranslated(const char* original) { return original; }
+consteval util::Untranslatable Untranslated(const char* original) { return original; }

 // Provide an overload of tinyformat::format which can take bilingual_str arguments.
 namespace tinyformat {
```

hodlinator commented:
Should this PR still be in Draft-status @maflcko?

maflcko reviewed (commented):
addressed feedback, rebased

hodlinator commented on src/util/translation.h:
```diff
@@ -47,13 +49,41 @@ inline bilingual_str operator+(bilingual_str lhs, const bilingual_str& rhs)
+template <unsigned num_params>
+struct bilingual_fmt {
+    const ConstevalFormatString<num_params> original;
+    consteval bilingual_fmt(Translatable o) : original{o.lit}
+    {
+    }
```
nit: Since previous review, this constructors function prototype has shrunk enough to keep it on one line, as you are doing for `Translatable` just above.
```suggestion
    consteval bilingual_fmt(Translatable o) : original{o.lit} {}
```

hodlinator commented on src/util/translation.h:
```diff
@@ -47,13 +49,41 @@ inline bilingual_str operator+(bilingual_str lhs, const bilingual_str& rhs)
+};
+
+template <unsigned num_params>
+struct bilingual_fmt {
```
nit: Use Consistent naming convention in newer code as recommended by developer-notes?
```suggestion
struct BilingualFmt {
```

hodlinator reviewed (changes requested):
Code reviewed fa6b1d2a26e25e3e6a4ef37835426371001d01ae

Really nice how you could slim this down after #31072 merged, removing `Translatable<false>` from this PR as after that PR `Untranslated()` is no longer sent into `strprintf`/`tfm::format`. Now we only need the two `Translatable` and `bilingual_fmt` as separate types as a consequence of all `_()`-calls not being used as format strings.

---

Only blocker for me is that the commit message in aaaa7e3f285d5d0ff55df4dfe050334215e2c582 still states:
> This is required for a future commit that requires Untranslated() and _() to be consteval for format literals.

`Untranslated()` is not made `consteval` in this PR.

Commit:
refactor: Introduce struct to hold a runtime format string

This brings the format types closer to the standard library types:

* FormatStringCheck corresponds to std::basic_format_string, with
  compile-time checks done via ConstevalFormatString
* RuntimeFormat corresponds to std::runtime_format, with no compile-time
  checks done.

Also, it documents where no compile-time checks are done.

Commit:
lint: Remove unused and broken format string linter

The linter has many implementation bugs and missing features.

Also, it is completely redundant with FormatStringCheck, which
constructs from ConstevalFormatString or a runtime format string.

hodlinator commented on src/tinyformat.h:
```diff
@@ -179,13 +179,19 @@ namespace tfm = tinyformat;
 
 namespace tinyformat {
 
+// Similar to std::runtime_format from C++26.
```
nit:
```diff
+// Added for Bitcoin Core.
```
Maybe use some comment style to group the modifications to avoid repeating this, similar to https://www.doxygen.nl/manual/grouping.html

hodlinator commented on src/tinyformat.h:
```diff
@@ -179,13 +179,19 @@ namespace tfm = tinyformat;
 struct FormatStringCheck {
     consteval FormatStringCheck(const char* str) : fmt{util::ConstevalFormatString<num_params>{str}.fmt} {}
-    FormatStringCheck(const std::string& str) : fmt{str.c_str()} {}
+    FormatStringCheck(const RuntimeFormat& run) : fmt{run.fmt.c_str()} {}
```
Just realized there's some risk of dangling pointers.
```suggestion
    FormatStringCheck(const RuntimeFormat& run LIFETIMEBOUND) : fmt{run.fmt.c_str()} {}
```

Also, if you're going to rename this one `str`->`run`, maybe call the other one `comp` or something?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1875730141

Delete seems good, but I think maybe LIFETIMEBOUND could still be helpful because `fmt` member here is being assigned the result of `run.fmt.c_str()`, so the `RuntimeFormat` object lifetime needs to be at least as long as `FormatStringCheck` object lifetime, and maybe LIFETIMEBOUND could help enforce this? I don't know all of the details about how LIFETIMEBOUND works though, so this is just a question.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1878803831

I'm still not sure I see any downsides to adding LIFETIMEBOUND. It seems like a great way to document the fact that these functions require long-lived references, and to provide instant feedback for code and design bugs without needing to create a PR and wait for an ASAN job to fail.

I also don't understand why the additional temporaries and moves and elisions in fad431e548a694600a3fe212127052e5f922a622 would preferable to using references and creating the minimum number of objects with straightforward full-expression lifetimes.

fad431e548a694600a3fe212127052e5f922a622 also seems unsafe because it doesn't prevent lvalues from being passed to the FormatStringCheck constructor, and if an lvalue (or any other argument incompatible with copy elision) is passed it will result in `FormatStringCheck::fmt` being assigned a dangling pointer.

Would suggest following changes as a simpler alternative:

```diff
--- a/src/tinyformat.h
+++ b/src/tinyformat.h
@@ -141,6 +141,7 @@ namespace tfm = tinyformat;
 
 //------------------------------------------------------------------------------
 // Implementation details.
+#include <attributes.h> // Added for Bitcoin Core
 #include <algorithm>
 #include <iostream>
 #include <sstream>
@@ -181,10 +182,8 @@ namespace tinyformat {
 
 // Added for Bitcoin Core. Similar to std::runtime_format from C++26.
 struct RuntimeFormat {
-    std::string fmt; // Not a string view, because tinyformat requires a c_str
-    explicit RuntimeFormat(std::string str) : fmt{std::move(str)} {}
-    RuntimeFormat(const RuntimeFormat&) = delete;
-    RuntimeFormat& operator=(const RuntimeFormat&) = delete;
+    const std::string& fmt; // Not a string view, because tinyformat requires a c_str
+    explicit RuntimeFormat(LIFETIMEBOUND const std::string& str) : fmt{str} {}
 };
 
 // Added for Bitcoin Core. Wrapper for checking format strings at compile time.
@@ -193,7 +192,7 @@ struct RuntimeFormat {
 template <unsigned num_params>
 struct FormatStringCheck {
     consteval FormatStringCheck(const char* str) : fmt{util::ConstevalFormatString<num_params>{str}.fmt} {}
-    FormatStringCheck(const RuntimeFormat& run) : fmt{run.fmt.c_str()} {}
+    FormatStringCheck(LIFETIMEBOUND const RuntimeFormat& run) : fmt{run.fmt.c_str()} {}
     FormatStringCheck(util::ConstevalFormatString<num_params> str) : fmt{str.fmt} {}
     operator const char*() { return fmt; }
     const char* fmt;
```

hodlinator replied:
`LIFETIMEBOUND` only being supported by one compiler isn't a great reason to skip adding it. Good to let Clang do what it can, and also provide documentation to devs. Not blocking for me though.

Tried the following, but Clang 18 compiles without any warnings when adding `LIFETIMEBOUND`. 
```C++
std::optional<tinyformat::FormatStringCheck<0>> fsc;
{
    auto fmt{std::make_unique<tinyformat::RuntimeFormat>("{}")};
    fsc = {tinyformat::FormatStringCheck<0>(*fmt)};
}
```

Also tried removing the `RuntimeFormat`-ctor from `FormatStringCheck` and found that it's only being used in `tinyformat::format(util::BilingualFmt...` outside off fuzz-tests. So this should be a minor concern.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1879761314

> Apart from the downsides I mentioned,
>
> It is just one reason, among many that have been mentioned previously.

Thanks for adding LIFETIMEBOUND here. I've been pretty confused by this discussion because I haven't seen any reasons given for why it would actually be better not to use LIFETIMEBOUND, and I don't know of any reasons. I understand LIFETIMEBOUND doesn't work everywhere, but don't understand why that matters as long as it catches bugs and provides fast feedback in the places it does work, and is helpful everywhere to document references that need to be long-lived.

If LIFETIMEBOUND triggered false positive errors like clang thread safety annotations do, that could be a real downside and reason not to use it. But as far as I know it doesn't cause harms or have any negative effects like that, so I've been confused by the pushback here. Maybe the reasons not to use it are as simple as "looks ugly"  / "adds noise"  / "seems useless" which are perfectly ok (and are also reasons I don't use `const` various places). i'd be fine with keeping or dropping LIFETIMEBOUND in this PR, just wondering if I am not understanding real problems with it.

ryanofsky replied:
Oh thanks for explaining that. I thought you were only disabling copies and assignment to prevent expensive copies due to not passing RuntimeFormat by reference, and that you were not passing RuntimeFormat by reference to avoid needing to think about lifetime issues.

But actually your goal is to make `RuntimeFormat run("fmt"); strprintf(run);` a compiler error, and deleting the copy and assignment operators are a way to do that. So LIFETIMEBOUND isn't harmful, but just not as good 
as a potentially safer solution fad431e548a694600a3fe212127052e5f922a622 that doesn't use LIFETIMEBOUND.

I think I can see now how fad431e548a694600a3fe212127052e5f922a622 actually is a safer solution in every case, and relying on mandatory copy elision is ok there because deleting the RuntimeFormat copy constructor should make it impossible to pass as a function argument without copy elision. This is more complicated than just using references, but does seem safer.

hodlinator replied:
>  https://godbolt.org/z/6336nGKcM (replace `0` with `1` to enable the compile-time checks)) into a compile failure on all compilers seems better than a runtime error, possibly only in CI.

With the `#if` set to `1` and lines 42-43 changed to
```C++
        check = RuntimeFormat("fmt");
```
One still gets an ASAN error instead of a compile time error.

hodlinator commented on src/tinyformat.h:
```diff
@@ -179,13 +179,19 @@ namespace tfm = tinyformat;
+// Similar to std::runtime_format from C++26.
+struct RuntimeFormat {
+    explicit RuntimeFormat(std::string str) : fmt{std::move(str)} {}
+    std::string fmt;
```
Noticed that the `std`-version uses `string_view`. Maybe add a note about why you didn't go with that in the commit message, or consider switching.

hodlinator reviewed (commented):
Thanks for incorporating my previous feedback!

Leaving some brief comments on the new changes before getting back to the weekend.

Was `lint-format-strings.py` picked up by some glob-expression in CI?

ryanofsky commented:
This looks pretty good but I think the API can be improved. Specifically I think `.translate()` suffixes should not be necessary and just calling `_()` should be sufficient to indicate that you want a translated `std::string` or `bilingual_str`. If you don't want the translation, there should be no reason to call `_()`

A suggested implementation dropping .translate() suffixes is: [branch](https://github.com/ryanofsky/bitcoin/commits/review.31061.9-edit), [diff](https://github.com/ryanofsky/bitcoin/compare/review.31061.9..review.31061.9-edit)

- daffb340d1d7b2c5a5248b3189a920cd3709e2ae refactor: Delay translation of _() literals
- 12e2e2bd21b3ff90df2afe91625ffcf930b0f48f refactor: Check translatable format strings at compile-time
- 3da69f27d1a0ca3c675f3cb491943beae9dc4922 refactor: Introduce struct to hold a runtime format string
- c357f990b71a7bd8b35abe08bd4ad309411a0325 lint: Remove unused and broken format string linter

This also adds a little more test coverage so the original and translated format strings in the test are not the same and test can confirm the right strings are picked.

l0rinc commented on src/util/translation.h:
```diff
@@ -47,13 +49,39 @@ inline bilingual_str operator+(bilingual_str lhs, const bilingual_str& rhs)
+consteval auto _(util::Translatable str) { return str; }
+
 /** Mark a bilingual_str as untranslated */
 inline bilingual_str Untranslated(std::string original) { return {original, original}; }
```
slightly unrelated: can we use a const reference here instead?
```suggestion
inline bilingual_str Untranslated(const std::string& original) { return {original, original}; }
```

hodlinator replied:
Undefined parameter evaluation order would make it unclear which of them should be explicitly moved though?
https://stackoverflow.com/a/2934909

I hope the compiler introduces an implicit move for whichever parameter it deems to be the last use of the variable.

ryanofsky reviewed (approved):
Code review ACK fa7cb525b7069ce0197d2f309102432273fea4d6. LGTM!

ryanofsky commented:
> Looks like there is a codegen bug in msvc, or some other issue? It would be good if someone with a Windows box could take a look, to either debug the crash or create a minimal reproducer.

I think maybe the previous fa37083b37e6 `FormatStringCheck(const RuntimeFormat& run)` was safe and current fad431e548a6 `FormatStringCheck(RuntimeFormat run)` is buggy and the msvc error is a real bug.

Previously it was safe to call `run.fmt.c_str()` because `run` object was temporary but had a lifetime was as long as the full-expression. Now it seems `run` object is a copy of original `RuntimeFormat` object and only has a lifetime as long as the FormatStringCheck constructor.

Not sure why this wouldn't cause errors on other platforms. One thing that may be unique about the timeoffsets test case that is failing is that the [format string](https://github.com/bitcoin/bitcoin/blob/37e49c2c7ca5969124830d79b2cb31041c570755/src/node/timeoffsets.cpp#L55-L62) is very long so maybe this is affected by a short string optimization.

ryanofsky commented:
> Maybe I am missing something obvious, but if there is no bug in msvc, it would be good to know the exact bug in this code.

I think the following line from fad431e548a694600a3fe212127052e5f922a622 is buggy because it is saving the result of `c_str()` call from a string that about to be destroyed:

```c++
FormatStringCheck(RuntimeFormat run) : fmt{run.fmt.c_str()} {}
```

The `run.fmt` string object will be destroyed when the `FormatStringCheck` constructor finishes because `run` is a non-reference parameter, so it's a effectively a local variable within `FormatStringCheck` method.

ryanofsky commented:
> I think the following line from https://github.com/bitcoin/bitcoin/commit/fad431e548a694600a3fe212127052e5f922a622 is buggy because it is saving the result of c_str() call from a string that about to be destroyed:

Looking at this more I think the constructor is not buggy in c++17 as long as a temporary object is being passed as the constructor argument, because the mandatory copy elision described at the top of https://en.cppreference.com/w/cpp/language/copy_elision should apply, so the temporary should not be copied, and the temporary should have the lifetime of the full-expression. So maybe MSVC is not applying mandatory copy-elision here, or not extending the lifetime of the constructor argument to outlive the constructor when copy-elision is applied.

I do think the FormatStringCheck constructor in fad431e548a694600a3fe212127052e5f922a622 is inherently fragile though, because it will be copying the constructor argument if a non-rvalue is passed, and calling .c_str() on the temporary copy that is about to be destroyed. It seems better to just use references and LIFETIMEBOUND as originally suggested and avoid temporaries and copies.

hodlinator commented:
I'd prefer it if every `TranslatedLiteral` instance didn't have a function pointer riding on the side, but I do like the added testability.

One could make the test-implementation of `G_TRANSLATION_FUN` support hooking it for specific tests. It still comes with some added overhead when it comes to benchmarks and fuzzing over current *master* though.

<details>
<summary>Diff</summary>

```diff
diff --git a/src/test/translation_tests.cpp b/src/test/translation_tests.cpp
index 2ebde95259..bcbfe9db34 100644
--- a/src/test/translation_tests.cpp
+++ b/src/test/translation_tests.cpp
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <test/util/setup_common.h>
 #include <tinyformat.h>
 #include <util/translation.h>
 
@@ -9,21 +10,24 @@
 
 BOOST_AUTO_TEST_SUITE(translation_tests)
 
-static TranslateFn translate{[](const char * str) {  return strprintf("t(%s)", str);  }};
-
 BOOST_AUTO_TEST_CASE(translation_namedparams)
 {
-    constexpr util::TranslatedLiteral format{"original [%s]", &translate};
+    assert(!G_TRANSLATION_FUN_HOOK);
+    G_TRANSLATION_FUN_HOOK = [](const char * str) { return strprintf("t(%s)", str); };
+
+    constexpr util::TranslatedLiteral format{"original [%s]"};
 
     bilingual_str arg{"original", "translated"};
     bilingual_str result{strprintf(format, arg)};
     BOOST_CHECK_EQUAL(result.original, "original [original]");
     BOOST_CHECK_EQUAL(result.translated, "t(original [translated])");
 
-    util::TranslatedLiteral arg2{"original", &translate};
+    util::TranslatedLiteral arg2{"original"};
     bilingual_str result2{strprintf(format, arg2)};
     BOOST_CHECK_EQUAL(result2.original, "original [original]");
     BOOST_CHECK_EQUAL(result2.translated, "t(original [t(original)])");
+
+    G_TRANSLATION_FUN_HOOK = {};
 }
 
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index 12feba09a5..947f1e1ca7 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -72,7 +72,11 @@ using node::LoadChainstate;
 using node::RegenerateCommitments;
 using node::VerifyLoadedChainstate;
 
-const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
+std::function<std::string(const char*)> G_TRANSLATION_FUN_HOOK;
+
+const std::function<std::string(const char*)> G_TRANSLATION_FUN{[] (const char* str) {
+    return G_TRANSLATION_FUN_HOOK ? G_TRANSLATION_FUN_HOOK(str) : str;
+}};
 
 constexpr inline auto TEST_DIR_PATH_ELEMENT{"test_common bitcoin"}; // Includes a space to catch possible path escape issues.
 /** Random context to get unique temp data dirs. Separate from m_rng, which can be seeded from a const env var */
diff --git a/src/test/util/setup_common.h b/src/test/util/setup_common.h
index b0f7bdade2..f033f20dc8 100644
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -43,6 +43,8 @@ extern const std::function<std::vector<const char*>()> G_TEST_COMMAND_LINE_ARGUM
 /** Retrieve the unit test name. */
 extern const std::function<std::string()> G_TEST_GET_FULL_NAME;
 
+extern std::function<std::string(const char*)> G_TRANSLATION_FUN_HOOK;
+
 static constexpr CAmount CENT{1000000};
 
 /** Register common test args. Shared across binaries that rely on the test framework. */
diff --git a/src/util/translation.h b/src/util/translation.h
index 3a9316479d..d618b22ee2 100644
--- a/src/util/translation.h
+++ b/src/util/translation.h
@@ -13,8 +13,7 @@
 #include <string>
 
 /** Translate a message to the native language of the user. */
-using TranslateFn = std::function<std::string(const char*)>;
-const extern TranslateFn G_TRANSLATION_FUN;
+const extern std::function<std::string(const char*)> G_TRANSLATION_FUN;
 
 /**
  * Bilingual messages:
@@ -54,10 +53,9 @@ namespace util {
 //! Compile-time literal string that can be translated with an optional translation function.
 struct TranslatedLiteral {
     const char* const original;
-    const TranslateFn* translate_fn;
 
-    consteval TranslatedLiteral(const char* str, const TranslateFn* fn = &G_TRANSLATION_FUN) : original{str}, translate_fn{fn} { assert(original); }
-    operator std::string() const { return translate_fn && *translate_fn ? (*translate_fn)(original) : original; }
+    consteval TranslatedLiteral(const char* str) : original{str} { assert(original); }
+    operator std::string() const { return G_TRANSLATION_FUN ? G_TRANSLATION_FUN(original) : original; }
     operator bilingual_str() const { return {original, std::string{*this}}; }
 };
 ```
</details>

waurah commented:
Есть телега?

ryanofsky commented on src/test/translation_tests.cpp:
```diff
@@ -9,13 +9,21 @@
+
 BOOST_AUTO_TEST_CASE(translation_namedparams)
 {
+    constexpr util::TranslatedLiteral format{"original [%s]", &translate};
```
In commit "refactor: Check translatable format strings at compile-time" (fa948257f332772f058cabffe0e313a0518da724)

Could consider making the test a little more realistic with:

```diff
--- a/src/test/translation_tests.cpp
+++ b/src/test/translation_tests.cpp
@@ -11,17 +11,22 @@ BOOST_AUTO_TEST_SUITE(translation_tests)
 
 static TranslateFn translate{[](const char * str) {  return strprintf("t(%s)", str);  }};
 
+// Custom translation function _t(), similar to _() but internal to this test.
+static inline consteval auto _t(util::TranslatedLiteral str)
+{
+    str.translate_fn = &translate;
+    return str;
+}
+
 BOOST_AUTO_TEST_CASE(translation_namedparams)
 {
-    constexpr util::TranslatedLiteral format{"original [%s]", &translate};
-
     bilingual_str arg{"original", "translated"};
-    bilingual_str result{strprintf(format, arg)};
+    bilingual_str result{strprintf(_t("original [%s]"), arg)};
     BOOST_CHECK_EQUAL(result.original, "original [original]");
     BOOST_CHECK_EQUAL(result.translated, "t(original [translated])");
 
     util::TranslatedLiteral arg2{"original", &translate};
-    bilingual_str result2{strprintf(format, arg2)};
+    bilingual_str result2{strprintf(_t("original [%s]"), arg2)};
     BOOST_CHECK_EQUAL(result2.original, "original [original]");
     BOOST_CHECK_EQUAL(result2.translated, "t(original [t(original)])");
 }
```

ryanofsky reviewed (approved):
Code review ACK fa7cb525b7069ce0197d2f309102432273fea4d6. Looks good but did leave some suggestions which could be saved for a follow-up.

---

re: https://github.com/bitcoin/bitcoin/pull/31061#issuecomment-2533132896

> I'd prefer it if every `TranslatedLiteral` instance didn't have a function pointer riding on the side, but I do like the added testability.

I don't think the `TranslatedLiteral::translate_fn` members should add overhead because TranslatedLiteral constructor is consteval, and in any case TranslatedLiteral objects are ephemeral, they exist only to call a translation function one time and then disappear. (The additional pointers also technically aren't function pointers, but pointers to a `std::function` object.)

The goal of the pointers isn't just improve testability but also to provide extensibility, so for example a libbitcoinkernel application could use define its own `_` translation function and seamlessly use strprintf with its own translated strings. Having the pointers also avoids the need to mutate shared global state during tests so, for example, unlike with the suggested change, other tests could run in parallel with the translation test without being affected, and if translation test throws, it will not leave G_TRANSLATION_FUN pointing at the wrong function.

hodlinator commented:
> I don't think the `TranslatedLiteral::translate_fn` members should add overhead because TranslatedLiteral constructor is consteval

Good point, I glossed over `TranslatedLiteral` being `consteval`.

> Having the pointers also avoids the need to mutate shared global state during tests

Current CTest executes tests in parallel through starting separate processes, but it might change at some point and keeping it less brittle is good.

hodlinator reviewed (approved):
re-ACK fa7cb525b7069ce0197d2f309102432273fea4d6

`git range-diff master fa6b1d2 fa7cb52`

Really like how `_()` is sufficient without `.translated`/`.translate()` thanks to `TranslatedLiteral`s `operator std::string`.

---

Checked out fa37083b37e613f0987fe74e0397d9ed90589e4b.

Ran linter being removed in following commit (fa7cb525b7069ce0197d2f309102432273fea4d6):
```
₿ test/lint/lint-format-strings.py 
src/test/fuzz/util/wallet.h: Expected 0 argument(s) after format string but found 2 argument(s): strprintf(tfm::RuntimeFormat{desc_fmt}, "[5aa9973a/66h/4h/2h]" + seed_insecure, int{internal})
src/test/translation_tests.cpp: Expected 0 argument(s) after format string but found 1 argument(s): strprintf(format, arg2)
```
Might want to do something about that if we want *all* commits to pass, not blocking for me though.

hodlinator commented:
nit: Noticed a typo in the linter removal commit message:

"The linter has many >implmentation< bugs and missing features."

ryanofsky reviewed (approved):
Code review ACK fadd2e69c6176ba684c1fd06fbd765be6fb6d647. Just rebased and added lifetimebound and test suggestions since last review. Appreciate your patience with all the review comments and iterations of this PR. Will be really nice to see this merged with the new compiler checks being safer and more convenient than current runtime errors.

hodlinator reviewed (approved):
re-ACK fadd2e69c6176ba684c1fd06fbd765be6fb6d647

`git range-diff master fa7cb52 fadd2e6`

- Mysteriously grinding out *fa*-prefixed hashes.
- Changed order of linter removal so that it doesn't error on what is now the last commit.
- Adapted unit tests further after feedback.
- Reconciled silent merge conflict with #30933.
- Added `LIFETIMEBOUND` to at least serve as developer documentation even though Clang isn't able to fully use it.

nit: Did not fix typo: https://github.com/bitcoin/bitcoin/pull/31061#issuecomment-2535546826

hodlinator reviewed (approved):
re-ACK fa6d9a29162fb508385201c926ec745d071086fc

`git range-diff master fadd2e6 fa6d9a2`

- Fixed typo in commit message
- Changed translate -> translated in lambda, ++correctness

ryanofsky reviewed (approved):
Code review ACK fa6d9a29162fb508385201c926ec745d071086fc. Just rebase, spelling fix and suggested rename since last review.

stickies-v commented on src/tinyformat.h:
```diff
@@ -179,13 +180,19 @@ namespace tfm = tinyformat;
+};
+
 // Added for Bitcoin Core. Wrapper for checking format strings at compile time.
 // Unlike ConstevalFormatString this supports std::string for runtime string
```
nit: `this supports std::string for runtime string` - perhaps worth updating to reflect that it now requires to be wrapped in `RuntimeFormat`?

stickies-v reviewed (commented):
Approach ACK fa6d9a29162fb508385201c926ec745d071086fc. Apologies for the slow re-review, but I like the latest form of this PR much more than the previous one I looked at, nice! I'll complete my full re-review in the next couple of days.

hodlinator reviewed (approved):
re-ACK fa51381790fe19e37e04a01a39cc94a00dcc441c

Just resolved conflict with #31616 since my last review.

ryanofsky reviewed (approved):
Code review ACK fa51381790fe19e37e04a01a39cc94a00dcc441c. Just rebased and fixed minor conflict since last review.

stickies-v reviewed (approved):
ACK fa51381790fe19e37e04a01a39cc94a00dcc441c - replacing linters with compile time checks is a very nice win. The newly introduced `TranslatedLiteral` and `BilingualFmt` make for a nice interface, and `RuntimeFormat` preserves the ability to bypass compile-time checks in a clear and convenient way. 

A couple of non-blocking nits (sorry, I thought I'd posted them already) that shouldn't hold up this PR.

Commit:
refactor: Use TranslateFn type consistently

The type was introduced in the previous commit.

stickies-v reviewed (approved):
re-ACK fa3efb5729091a36a0e82316e9e4b7c09115dc2e

No changes except for consistent re-use of `TranslateFn`, minor docstring change and un-inline-ing test translate function.

ryanofsky reviewed (approved):
Code review ACK fa3efb5729091a36a0e82316e9e4b7c09115dc2e. Since last review added TranslateFn commit, clarified FormatStringCheck documentation, dropped redundant `inline` keyword
